### ISSUE TITLE
k3s-root binaries cause segmentation fault on aarch64 nodes with 64k page size

### ISSUE THREAD

**Issue by leoluo0818** [k3s-io/k3s#6708]
**Environmental Info:**
k3s version v1.23.15+k3s1 (50cab3b3)
go version go1.17.13

Node(s) CPU architecture, OS, and Version: 

Linux private.local 4.18.0-147.8.1.el7.aarch64 #1 SMP Wed Apr 15 18:13:44 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux

Cluster Configuration:

3 master, 1 worker

**Describe the bug:**

1、show by `kubectl get pods -A` after `kubectl apply`
```
mynamespace           nginx-5785485ddf-w4v7p                    0/1     ImagePullBackOff   0             1m
```
2、kubectl describe pod <podname>:
```
Failed to pull image "registry.private.org/nginx:latest": rpc error: code = Unknown desc = failed to pull and unpack image "registry.private.org/nginx:latest": failed to extract layer sha256:4ca5bcf319c0635e3c47b183c1e28f3f79acd6bc22e91fc4102bd5798db73f24: mount callback failed on /var/lib/rancher/k3s/agent/containerd/tmpmounts/containerd-mount3910203887: signal: segmentation fault: : unknown
```

3、 k3s ctr image pull registry.private.org/nginx:latest --debug:

```
registry.private.org/nginx:latest:                        resolved       |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:add49c062760eeb91159d9adad8c9f281a680eefa66bf3f8f26500b1ddd85a49: exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:c453544db8bbc074b616d504ad43417741e74050e844af2c24827316ed1df0f8:    exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:5698c81efc1fc84f028f2ee239309f7d48543d6a7454242b04a5034e249f908c:    exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:b81bff4db76116e119103029bb5d3eabb22bc21f2dcf7fc640867b516692a1f0:    exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:78d8281b4b28bab3a16c4fd4069f05b43979826d042ce43dc4743315ee3e3a7e:    exists         |++++++++++++++++++++++++++++++++++++++|
config-sha256:1868dce659f5a83b3ccd4468c780157dbb56d6c5eaacd4bcabd951378882d195:   exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:ec1dea3d047340fd4d4fcb25962ff149e63578f83ac540d2b98179bc2425bcac:    exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:9b18e9b68314027565b90ff6189d65942c0f7986da80df008b8431276885218e:    exists         |++++++++++++++++++++++++++++++++++++++|
elapsed: 0.2 s                                                                    total:   0.0 B (0.0 B/s)
unpacking linux/arm64/v8 sha256:add49c062760eeb91159d9adad8c9f281a680eefa66bf3f8f26500b1ddd85a49...

INFO[0000] apply failure, attempting cleanup             error="failed to extract layer sha256:5d3e392a13a0fdfbf8806cb4a5e4b0a92b5021103a146249d8a2c999f06a9772: mount callback failed on /var/lib/rancher/k3s/agent/containerd/tmpmounts/containerd-mount3909800577: signal: segmentation fault: : unknown" key="extract-813510370-nYc3 sha256:5d3e392a13a0fdfbf8806cb4a5e4b0a92b5021103a146249d8a2c999f06a9772"
ctr: failed to extract layer sha256:5d3e392a13a0fdfbf8806cb4a5e4b0a92b5021103a146249d8a2c999f06a9772: mount callback failed on /var/lib/rancher/k3s/agent/containerd/tmpmounts/containerd-mount3909800577: signal: segmentation fault: : unknown
```
4、 k3s ctr image list | grep nginx
```
registry.private.org/nginx:latest               application/vnd.docker.distribution.manifest.v2+json sha256:<holder> 142.9 MiB linux/arm64 io.cri-containerd.image=managed

```

**Steps To Reproduce:**

- Installed K3s: 
```
autok3s create \
  --provider native \
  --cluster \
  --enable  dashboard \
  --name  custom-staging \
  --package-name  custom \
  --ssh-key-path  ./devops_rsa \
  --ssh-port  22 \
  --ssh-user  root \
  --master-ips  10.2.178.51,10.2.178.52,10.2.178.53 \
  --worker-ips 10.2.178.54 \
  --registry ./registries.yaml

```
**Expected behavior:**

**Actual behavior:**

**Additional context / logs:**

**Comment by brandond** [comment]
`mount callback failed on /var/lib/rancher/k3s/agent/containerd/tmpmounts/containerd-mount3909800577: signal: segmentation fault: : unknown`

I've not seen this before, but it sounds like the overlayfs mount command is segfaulting? Do you see anything in the kernel message log, or in the containerd log file?

**Comment by wenerme** [comment]
```yaml
OS: CentOS Linux release 7.9.2009 (AltArch) aarch64
Host: Alibaba Cloud ECS virt-rhel7.6.0
Kernel: 4.18.0-348.20.1.el7.aarch64
CPU: (16)
GPU: Red Hat, Inc. Virtio GPU
```

```
E0209 23:44:50.577685   22502 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"CreatePodSandbox\" for \"metrics-server-5f9f776df5-c24mt_kube-system(8f8bcf31-e1bf-47fc-a1a7-9b3dd7fbebef)\" with CreatePodSandboxError: \"Failed to create sandbox for pod \\\"metrics-server-5f9f776df5-c24mt_kube-system(8f8bcf31-e1bf-47fc-a1a7-9b3dd7fbebef)\\\": rpc error: code = Unknown desc = failed to get sandbox image \\\"rancher/mirrored-pause:3.6\\\": failed to pull image \\\"rancher/mirrored-pause:3.6\\\": failed to pull and unpack image \\\"docker.io/rancher/mirrored-pause:3.6\\\": failed to extract layer sha256:c640e628658788773e4478ae837822c9bc7db5b512442f54286a98ad50f88fd4: mount callback failed on /data/k3s/agent/containerd/tmpmounts/containerd-mount1211601772: signal: segmentation fault: : unknown\"" pod="kube-system/metrics-server-5f9f776df5-c24mt" podUID=8f8bcf31-e1bf-47fc-a1a7-9b3dd7fbebef
```

```
k3s version v1.25.6+k3s1 (9176e03c)
go version go1.19.5
```

This is aliyun provided el7 arm64

dmesg 

```
[111874.143252] IPVS: Registered protocols (TCP, UDP, SCTP, AH, ESP)
[111874.144745] IPVS: Connection hash table configured (size=4096, memory=64Kbytes)
[111874.146099] IPVS: ipvs loaded.
[111874.151381] IPVS: [rr] scheduler registered.
[111874.155030] IPVS: [wrr] scheduler registered.
[111874.158517] IPVS: [sh] scheduler registered.
[112323.806691] cgroup: cgroup: disabling cgroup2 socket matching due to net_prio or net_cls activation
```

config

```yaml
write-kubeconfig-mode: '0644'

data-dir: /data/k3s

token: 
agent-token: 

node-name: svr-1

disable:
- traefik
```

**Comment by brandond** [comment]
Can you post the output of `cat /proc/cpuinfo` on these nodes?

Do you see the same error on this hardware if using an OS other than EL7, or with a newer kernel?

**Comment by wenerme** [comment]
```bahs
cat /proc/cpuinfo
```

```txt
processor	: 15
BogoMIPS	: 50.00
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x3
CPU part	: 0xd0c
CPU revision	: 1
```

**Comment by brandond** [comment]
Is that all of it? Can you pos the complete output?

**Comment by wenerme** [comment]
They are. same

```bash
cat /proc/cpuinfo | sort -u
```

```
BogoMIPS	: 50.00
CPU architecture: 8
CPU implementer	: 0x41
CPU part	: 0xd0c
CPU revision	: 1
CPU variant	: 0x3
Features	: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
processor	: 0
processor	: 1
processor	: 10
processor	: 11
processor	: 12
processor	: 13
processor	: 14
processor	: 15
processor	: 2
processor	: 3
processor	: 4
processor	: 5
processor	: 6
processor	: 7
processor	: 8
processor	: 9
```

> Do you see the same error on this hardware if using an OS other than EL7, or with a newer kernel?

No error in dmesg, can not change os or kernel, not in my control, sorry. Currently using k3d as workaround.

**Comment by brandond** [comment]
Hmm. That's a lot sparser than you get on most platforms. A bit of googling suggests this a Graviton 2 processor, is that correct? Unfortunately I don't have access to test on Alibaba Cloud, but I have run K3s on AWS Graviton 2 nodes without issue.

I find it odd that k3d works, as you end up running k3s with the same userspace binaries in the end. Are you able to reproduce this on other Linux distributions on the same hardware?

**Comment by gdoucet** [comment]
I got the same issue on Oracle Cloud Arm vm:
# lscpu 
Architecture:        aarch64
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per cluster: 4
Socket(s):           1
Cluster(s):          1
NUMA node(s):        1
Vendor ID:           ARM
BIOS Vendor ID:      QEMU
Model:               1
Model name:          Neoverse-N1
BIOS Model name:     virt-4.2
Stepping:            r3p1
BogoMIPS:            50.00
NUMA node0 CPU(s):   0-3
Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
# cat /proc/cpuinfo 
processor       : 0
BogoMIPS        : 50.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x3
CPU part        : 0xd0c
CPU revision    : 1

processor       : 1
BogoMIPS        : 50.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x3
CPU part        : 0xd0c
CPU revision    : 1

processor       : 2
BogoMIPS        : 50.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x3
CPU part        : 0xd0c
CPU revision    : 1

processor       : 3
BogoMIPS        : 50.00
Features        : fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
CPU implementer : 0x41
CPU architecture: 8
CPU variant     : 0x3
CPU part        : 0xd0c
CPU revision    : 1

**Comment by wenerme** [comment]
> Hmm. That's a lot sparser than you get on most platforms. A bit of googling suggests this a Graviton 2 processor, is that correct? Unfortunately I don't have access to test on Alibaba Cloud, but I have run K3s on AWS Graviton 2 nodes without issue.
> 
> I find it odd that k3d works, as you end up running k3s with the same userspace binaries in the end. Are you able to reproduce this on other Linux distributions on the same hardware?

K3s is ok, the problem is bundled coreutils

**Comment by brandond** [comment]
> K3s is ok, the problem is bundled coreutils

Yes, but any given release of k3s bundles the same coreutils in both the multicall binary, and in the docker image - so I'm at a bit if as loss as to why it works in k3d on the same hardware. Other than just updating the buildroot version, we haven't really changed much recently (ref: https://github.com/k3s-io/k3s-root/commit/48b49ad763ed1a6ca69b4cd51b6aa671cfdb97d8#diff-4579dd704caea98a79d18bc4b5f8c6bacfd6a5dae74f9c96f3506422d11a197b)

Would it be possible to get a core dump from the faulting binary on one of the affected nodes?

**Comment by gdoucet** [comment]
I tried to create a core dump but the process don't even start. Here is a session with gdb:

```
# gdb /var/lib/rancher/k3s/data/current/bin/coreutils 
GNU gdb (GDB) Red Hat Enterprise Linux 8.2-19.0.1.el8
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /var/lib/rancher/k3s/data/current/bin/coreutils...(no debugging symbols found)...done.
(gdb) catch throw
Catchpoint 1 (throw)
(gdb) break main
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 2 (main) pending.
(gdb) run
Starting program: /var/lib/rancher/k3s/data/9088e57b1ba3c37820aaba60202af921dbc01b77ec0ad1e08be86b5c7bc9b8c1/bin/coreutils 
During startup program terminated with signal SIGSEGV, Segmentation fault.
```

**Comment by brandond** [comment]
Hmm, is it just coreutils or does busybox also segfault?

We haven't changed our target core, we're still building for `BR2_cortex_a53` which should be forwards-compatible, but I wonder if there's some specific CPU instruction that the compiler isn't handling properly.

**Comment by gdoucet** [comment]
Only coreutils busybox works fine.

**Comment by gdoucet** [comment]
I rebuilded k3s-root with:
```
BR2_ARM64_PAGE_SIZE_64K=y
BR2_ARM64_PAGE_SIZE="64K"
```
And there is no segfault with coreutils.

It had segfault with:
```
BR2_ARM64_PAGE_SIZE_4K=y
BR2_ARM64_PAGE_SIZE="4K"
```

**Comment by brandond** [comment]
Hmm, that's interesting... unfortunately not compatible with other hardware.

**Comment by brandond** [comment]
I guess this is actually a kernel compatibility option; I am surprised that it is having this effect on userspace binaries. Are you able to switch back to 4k page sizes?

I can see what happens if we switch to 64k but I am concerned it will break compatibility with other platforms like Raspberry Pi and AWS where K3s is currently functional.

**Comment by gdoucet** [comment]
I am puzzled why the binaries compiled with golang are not affected, only the one from k3s-root. I will try to see if there is a kernel setting I can change.

**Comment by gdoucet** [comment]
Seen to be only affecting Centos 8 / Oracle Linux 8. I tried with Oracle Linux 9 and there is no segfault.

**Comment by brandond** [comment]
EL9 has gone back to 4k pages for this very reason: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html-single/9.0_release_notes/index
> Red Hat has selected a 4 KB page size of physical memory for the 64-bit ARM architecture in Red Hat Enterprise Linux 9. This size pairs well with the workloads and memory amounts present on the majority of ARM-based systems. To employ large page sizes efficiently, use the huge pages option to address a greater amount of memory or workloads with large data sets.

SLE does the same: https://www.suse.com/releasenotes/x86_64/SUSE-SLES/15-SP4/index.html#technology-preview-aarch64
> SUSE Linux Enterprise Server for Arm 12 SP2 and later kernels have used a page size of 4K. This offers the widest compatibility also for small systems with little RAM, allowing to use Transparent Huge Pages (THP) where large pages make sense.
> As a technology preview, SUSE Linux Enterprise Server for Arm 15 SP3 added a kernel flavor 64kb, offering a page size of 64 KiB and physical/virtual address size of 52 bits.

At this point I am leaning towards documenting that k3s requires a kernel 4k page size on aarch64.

**Comment by wgbeckmann** [comment]
Hi, I'm not shure of this is a joke. 
As I said I'm in an VM on the Oracle Cloud. How can I Switch the Kernel?

**Comment by brandond** [comment]
I would consult Oracle support for assistance with that. We don't have any Oracle cloud experts on our team, nor do we have access to it for testing.

**Comment by dereknola** [comment]
Closing this as `won't fix`. K3s is not going to be built to support aarch64 with non-4k page sizes. We have updated the documentation to call this out. 

https://docs.k3s.io/installation/requirements#architecture
